### PR TITLE
feat(braze): nested array ops

### PR DESCRIPTION
**Pull request by aashishmalik**
## Description of the change

> Braze: Update cloud mode implementation to reduce data point use

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/braze/transform.js [hunk 1]
```diff
@@ -308,6 +309,37 @@ function processTrackEvent(messageType, message, destination, mappingJson) {
   payload = addMandatoryEventProperties(payload, message);
   payload.properties = properties;
 
+  try {
+    // add,update,remove
+    if (destination.Config.enableNestedArrayOperations) {
+      Object.keys(properties)
+        .filter(k => Array.isArray(properties[`${k}`]))
```
Comment (ItsSudip) [inline src/v0/destinations/braze/transform.js:316]:
```suggestion
        .filter(key => Array.isArray(properties[`${key}`]))
```

File: src/v0/destinations/braze/transform.js [hunk 2]
```diff
@@ -308,6 +309,37 @@ function processTrackEvent(messageType, message, destination, mappingJson) {
...
+              )
+                ? properties.mergeObjectsUpdateOperation
+                : true;
+            }
+          }
+        });
+      delete properties.nestedOperationType;
+    }
+  } catch (exp) {
+    logger.info("Failure occured during nested array operations", exp);
```
Comment (ItsSudip) [inline src/v0/destinations/braze/transform.js:340]:
```suggestion
    logger.info("Failure occured during nested array operations", exp);
```
Logger is not defined.

File: test/__tests__/data/braze_input.json [hunk 3]
```diff
@@ -5,7 +5,8 @@
         "restApiKey": "9432f11f70f8ce386f5110c8c924b3ec4f825256",
         "prefixProperties": true,
         "useNativeSDK": false,
-        "dataCenter": "us-01"
+        "dataCenter": "us-01",
+        "enableNestedArrayOperations": true
```
Comment (ItsSudip) [inline test/__tests__/data/braze_input.json:9]: No need to add it with old test cases.

File: src/v0/destinations/braze/config.js [hunk 4]
```diff
@@ -32,6 +32,18 @@ const IDENTIFY_BRAZE_MAX_REQ_COUNT = 50;
...
+const nestedOpertationTypes = {
+  REMOVE: "remove",
+  UPDATE: "update",
+  ADD: "add"
+};
+
+const supportedOperationTypes = [
+  nestedOpertationTypes.ADD,
+  nestedOpertationTypes.REMOVE,
+  nestedOpertationTypes.UPDATE
```
Comment (ItsSudip) [inline src/v0/destinations/braze/config.js:44]:
```suggestion
  nestedOperationTypes.ADD,
  nestedOperationTypes.REMOVE,
  nestedOperationTypes.UPDATE
```

File: src/v0/destinations/braze/config.js [hunk 5]
```diff
@@ -32,6 +32,18 @@ const IDENTIFY_BRAZE_MAX_REQ_COUNT = 50;
 
 const DESTINATION = "braze";
 
+const nestedOpertationTypes = {
```
Comment (ItsSudip) [inline src/v0/destinations/braze/config.js:35]:
```suggestion
const nestedOperationTypes = {
```

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: test/__tests__/data/braze_input.json (hunk 3 above)
Comment (aashishmalik) [reply test/__tests__/data/braze_input.json:9]: added it to keep previous flow consistent

File: src/v0/destinations/braze/transform.js (hunk 1 above)
Comment (aashishmalik) [reply src/v0/destinations/braze/transform.js:319]: using 'key' later , I think its fine to use 'k' for this filter block

Comment (sandeepdsvs) [reply src/v0/destinations/braze/transform.js:319]: You can use key at both places

File: src/v0/destinations/braze/transform.js [hunk 6]
```diff
@@ -308,10 +312,82 @@ function processTrackEvent(messageType, message, destination, mappingJson) {
...
   if (payload) {
     requestJson.events = [payload];
   }
+
+  // update payload as per https://www.braze.com/docs/user_guide/data_and_analytics/custom_data/custom_attributes/array_of_objects/#api-request-body
+  if (
+    destination.Config.enableNestedArrayOperations &&
+    isDefinedAndNotNull(properties.nestedOperationType)
+  ) {
+    delete requestJson.events;
```
Comment (yashasvibajpai) [inline src/v0/destinations/braze/transform.js:384]: We should ideally avoid explicit deletion of object properties in our transformation code. Can we utilize something like [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) if possible? CC @ItsSudip 

|non blocking|

Comment (aashishmalik) [reply src/v0/destinations/braze/transform.js:384]: As this is final payload after transformation, I think we can delete un-required keys from it. These props are req for other flow, not for this change.

**Review by sandeepdsvs — APPROVED** [review]

File: test/__tests__/data/braze_input.json (hunk 3 above)
Comment (ItsSudip) [reply test/__tests__/data/braze_input.json:9]: If you do not provide the new variable it should be consistent otherwise it is not backward compatible

Comment (aashishmalik) [reply test/__tests__/data/braze_input.json:9]: this UT is showing - flag is enabled in UI, but we are not providing fields to use this flow in payload. So, old flow should work as previous.

**Review by ItsSudip — APPROVED** [review]

**Review by sandeepdsvs — APPROVED** [review]

**Review by yashasvibajpai — APPROVED** [review]

**Pull request merged**